File: 20221029_Saturday/20221029.go
// Leetcode: 2136. Earliest Possible Day of Full Bloom.
// https://leetcode.com/problems/earliest-possible-day-of-full-bloom/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 283 ms, faster than 92.86% of Go online submissions for Earliest Possible Day of Full Bloom.
// Memory Usage: 10.3 MB, less than 100.00% of Go online submissions for Earliest Possible Day of Full Bloom.
// 2022.10.29 Daily Challenge.
import "sort"
func earliestFullBloom(plantTime []int, growTime []int) int {
  // Based on:
  // https://leetcode.com/problems/earliest-possible-day-of-full-bloom/solution/
  current_plant_time := 0
  result := 0
  indices := make([]int, len(growTime))
  for i,_ := range indices {
    indices[i] = i
  }
  sort.SliceStable(indices,func (i,j int) bool{
    // Here was a bug. Fixed. i and j are incices of array.
    // and in array indices need to use value of this array 
    // in this case.
    return growTime[indices[i]] > growTime[indices[j]]
  })
  for _, i := range indices{
    current_plant_time += plantTime[i]
    result = max(result, current_plant_time + growTime[i])
  }
  return result
}
func max(a, b int) int {
  if a > b {
    return a
  } else {
    return b
  }
}

File: 20221029_Saturday/20221029.rb
# Leetcode: 2136. Earliest Possible Day of Full Bloom.
# https://leetcode.com/problems/earliest-possible-day-of-full-bloom/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 1249 ms, faster than 50.00% of Ruby online submissions for Earliest Possible Day of Full Bloom.
# Memory Usage: 227.8 MB, less than 50.00% of Ruby online submissions for Earliest Possible Day of Full Bloom.
# 2022.10.29 Daily Challenge.
# @param {Integer[]} plant_time
# @param {Integer[]} grow_time
# @return {Integer}
def earliest_full_bloom(plant_time, grow_time)
  # Based on:
  # https://leetcode.com/problems/earliest-possible-day-of-full-bloom/solution/
  cur_plant_time = 0
  result = 0
  indices = grow_time.map.with_index{|x,i| [x,i]}.sort_by{|x| -x[0]}.map(&:last)
  indices.each do |i|
    cur_plant_time += plant_time[i]
    result = [result, cur_plant_time + grow_time[i]].max
  end
  return result
end
